File: c/2.两数相加.c
/*
 * @lc app=leetcode.cn id=2 lang=c
 *
 * [2] 两数相加
 *
 * https://leetcode-cn.com/problems/add-two-numbers/description/
 *
 * algorithms
 * Medium (36.18%)
 * Likes:    3585
 * Dislikes: 0
 * Total Accepted:    281.5K
 * Total Submissions: 777.6K
 * Testcase Example:  '[2,4,3]\n[5,6,4]'
 *
 * 给出两个 非空 的链表用来表示两个非负的整数。其中，它们各自的位数是按照 逆序 的方式存储的，并且它们的每个节点只能存储 一位 数字。
 * 
 * 如果，我们将这两个数相加起来，则会返回一个新的链表来表示它们的和。
 * 
 * 您可以假设除了数字 0 之外，这两个数都不会以 0 开头。
 * 
 * 示例：
 * 
 * 输入：(2 -> 4 -> 3) + (5 -> 6 -> 4)
 * 输出：7 -> 0 -> 8
 * 原因：342 + 465 = 807
 * 
 * 
 */

// @lc code=start

#include <stdio.h>
#include <stdlib.h>

//Definition for singly-linked list.
// struct ListNode
// {
//     int val;
//     struct ListNode *next;
// };

struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2)
{
    struct ListNode *res, *p;
    res = NULL;
    p = NULL;
    int dec = 0;
    while (l1 || l2)
    {
        if (p == NULL)
        {
            res = (struct ListNode *)malloc(sizeof(struct ListNode));
            p = res;
        }
        else
        {
            p->next = (struct ListNode *)malloc(sizeof(struct ListNode));
            p = p->next;
        }

        p->val = dec;
        p->next = NULL;

        if (l1)
        {
            p->val += l1->val;
            l1 = l1->next;
        }
        if (l2)
        {
            p->val += l2->val;
            l2 = l2->next;
        }

        dec = 1 ? p->val >= 10 : 0; //下次进位
        p->val = p->val % 10;       //本次最终值
    }
    if (dec == 1)
    {
        p->next = (struct ListNode *)malloc(sizeof(struct ListNode));
        p->next->val = 1;
        p->next->next = NULL;
    }
    return res;
}

// @lc code=end

File: python/2.两数相加.py
#
# @lc app=leetcode.cn id=2 lang=python3
#
# [2] 两数相加
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        res = ListNode(-1)
        res_loop = res
        remind = 0
        while (l1 or l2):
            value = 0
            if l1:
                value = value + l1.val
                l1 = l1.next
            if l2:
                value = value + l2.val
                l2 = l2.next
            value = value + remind
            remind = int(value / 10)
            # create a new node.
            tmpL = ListNode(value % 10)
            res_loop.next = tmpL
            res_loop = res_loop.next
        if remind:
            res_loop.next = ListNode(remind)

        return res.next

"""
def printL(l):
    print('')
    while (l):
        print(f"-->{l.val}",end='')
        l = l.next
    print('')


test         = Solution()
l1           = ListNode(2)
l1.next      = ListNode(4)
l1.next.next = ListNode(3)
l2           = ListNode(5)
l2.next      = ListNode(6)
l2.next.next = ListNode(4)
printL(l1)
printL(l2)
l3 = test.addTwoNumbers(l1, l2)
printL(l3)

"""
# @lc code=end
